### PR TITLE
chore: remove redundant --rsc-pro install generator flag

### PR DESCRIPTION
**Pull request by ihabadham**
## Summary

#2821 introduced a dedicated `--rsc-pro` install flag. It was closed the next
day with: "Closing as redundant: current generator semantics already treat --rsc
as including Pro by design, so introducing --rsc-pro adds an unnecessary parallel
mode."

The flag then shipped anyway in #2822, which was primarily about the Pro upgrade
automation (#2626 — Gemfile swap, import rewriter). The --rsc-pro bits rode along
as a ~10% passenger in that larger PR and were never removed after #2821's close.

#3098 was then opened to add more tests for the flag, further entrenching
something that was explicitly called redundant.

This PR completes the intent behind #2821's close: removes the flag and folds
its useful behaviors into --rsc, where they belong.

## What moves from --rsc-pro to --rsc

- Pro verification checklist now fires on any --rsc install
- Prerelease installation note now fires on any --rsc install with a prerelease
  gem version
- Exact version pin for the Pro gem now triggers on
  `Gem::Version.new(ReactOnRails::VERSION).prerelease?` instead of the flag —
  this is the real reason exact pinning matters (Bundler's `~>` doesn't match
  prereleases), not a flag-specific behavior

Side effect: plain `--rsc` on a prerelease gem version now installs correctly
where it previously would have failed against `~>` pinning.

## What is removed

- `class_option :rsc_pro`
- `use_rsc_pro_mode?` helper
- `--rsc-pro` branches in `use_pro?`, `use_rsc?`, `pro_requirement_flag`,
  `recovery_install_command`
- All `--rsc-pro` test contexts

## What is NOT touched

`pro_generator.rb` and `pro_generator_spec.rb` (the #2626 upgrade automation —
Gemfile swap, import rewriter) are independent of this flag and untouched. Bugs
in that code are tracked separately in #3104.

## Closes / supersedes

Supersedes #3098 (its test assertions were inside `context "with --rsc-pro"`,
which no longer exists).

## Test plan

- `bundle exec rubocop` on all modified files — clean
- CI to run the full rspec suite

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Removed**
  * The Pro-specific install flag `--rsc-pro` has been removed.

* **Changed**
  * Pro-related behaviors that previously required `--rsc-pro` now apply when using `--rsc` (post-install verification checklist, prerelease install notes, and exact Pro gem pinning for prereleases).

* **Chore**
  * Added a URL exclusion to link-checking configuration to skip a specific domain.

### CODE REVIEW

File: react_on_rails/lib/generators/react_on_rails/pro_setup.rb [hunk 1]
```diff
@@ -94,20 +94,17 @@ def pro_flag_specified_for_context?
...
         return "--rsc" if options[:rsc]
 
         "--pro"
       end
 
-      def rsc_pro_prerelease_note
-        return "" unless use_rsc_pro_mode?
-        return "" unless Gem::Version.new(ReactOnRails::VERSION).prerelease?
+      def rsc_prerelease_note
+        return "" unless use_rsc?
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply react_on_rails/lib/generators/react_on_rails/pro_setup.rb]: Addressed in f867f9433. Renamed `rsc_prerelease_note` → `prerelease_note` and swapped the gate from `use_rsc?` to `prerelease_ror_version?` — the same condition that triggers the exact pin in `pro_gem_version_requirement`. `--pro` users on prerelease gems now get both the exact pin and the explanatory note.

[the remark this answers is not in the document]
Comment (ihabadham) [reply react_on_rails/lib/generators/react_on_rails/pro_setup.rb]: Addressed in f867f9433. Renamed `rsc_prerelease_note` → `prerelease_note` and swapped the gate from `use_rsc?` to `prerelease_ror_version?` — the same condition that triggers the exact pin in `pro_gem_version_requirement`. `--pro` users on prerelease gems now get both the exact pin and the explanatory note.

[the remark this answers is not in the document]
Comment (ihabadham) [reply react_on_rails/lib/generators/react_on_rails/pro_setup.rb]: Addressed in f867f9433. Renamed `rsc_prerelease_note` → `prerelease_note` and swapped the gate from `use_rsc?` to `prerelease_ror_version?` — the same condition that triggers the exact pin in `pro_gem_version_requirement`. `--pro` users on prerelease gems now get both the exact pin and the explanatory note.

File: react_on_rails/lib/generators/react_on_rails/install_generator.rb [hunk 2]
```diff
@@ -489,7 +481,7 @@ def add_post_install_message
                                      shakapacker_just_installed: shakapacker_just_installed?,
                                      landing_page: options.new_app? && new_app_root_route_available?
                                    ))
-        GeneratorMessages.add_info(rsc_pro_verification_message) if use_rsc_pro_mode?
+        GeneratorMessages.add_info(rsc_pro_verification_message) if use_rsc?
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply react_on_rails/lib/generators/react_on_rails/install_generator.rb]: Addressed in f867f9433. Renamed the method and its call site to `rsc_verification_message`. Kept the heading text ("🔎 RSC Pro Verification") since `--rsc` implies Pro and the content description is still accurate — only the identifier changed.

File: react_on_rails/lib/generators/react_on_rails/install_generator.rb [hunk 3]
```diff
@@ -644,7 +636,7 @@ def add_post_install_message
                                      ci_workflow_generated: @ci_workflow_generated == true,
                                      app_root: destination_root
                                    ))
-        GeneratorMessages.add_info(rsc_pro_verification_message) if use_rsc_pro_mode?
+        GeneratorMessages.add_info(rsc_pro_verification_message) if use_rsc?
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply react_on_rails/lib/generators/react_on_rails/install_generator.rb]: Addressed in f867f9433. Renamed the method and its call site to `rsc_verification_message`. Kept the heading text ("🔎 RSC Pro Verification") since `--rsc` implies Pro and the content description is still accurate — only the identifier changed.

File: react_on_rails/lib/generators/react_on_rails/pro_setup.rb [hunk 4]
```diff
@@ -94,20 +94,17 @@ def pro_flag_specified_for_context?
...
+        return "" unless prerelease_ror_version?
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply react_on_rails/lib/generators/react_on_rails/pro_setup.rb]: Addressed in f867f9433. Renamed `rsc_prerelease_note` → `prerelease_note` and swapped the gate from `use_rsc?` to `prerelease_ror_version?` — the same condition that triggers the exact pin in `pro_gem_version_requirement`. `--pro` users on prerelease gems now get both the exact pin and the explanatory note.

File: CHANGELOG.md [hunk 5]
```diff
@@ -24,6 +24,10 @@ After a release, run `/update-changelog` in Claude Code to analyze commits, writ
 
 ### [Unreleased]
 
+#### Removed
+
+- **[Pro]** **Removed the `--rsc-pro` install generator flag**: The flag was redundant with `--rsc` (which already implies Pro) and was originally closed as an "unnecessary parallel mode" in [PR 2821](https://github.com/shakacode/react_on_rails/pull/2821) before shipping unintentionally alongside the Pro upgrade automation in [PR 2822](https://github.com/shakacode/react_on_rails/pull/2822). The useful behaviors previously gated on `--rsc-pro` now fire on any `--rsc` install: the Pro verification checklist after install, the prerelease installation note when running on a prerelease gem version, and exact version pinning for the Pro gem when `ReactOnRails::VERSION` is a prerelease (since Bundler's `~>` does not match prerelease versions). Stable `--rsc` installs continue to use `~>` pinning. [PR 3104](https://github.com/shakacode/react_on_rails/issues/3104) tracks unrelated silent-failure bugs in the remaining upgrade automation. [ihabadham](https://github.com/ihabadham).
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply CHANGELOG.md]: Declining the URL change — #3104 is an issue (filed via `gh issue create` earlier in this cycle), so `/issues/3104` is the correct URL. The actual bug was the label — `[PR 3104]` was a typo pointing at an issue URL. Fixed in f867f9433 by relabeling to `[Issue 3104]` while keeping the URL unchanged.

File: CHANGELOG.md [hunk 6]
```diff
@@ -24,6 +24,10 @@ After a release, run `/update-changelog` in Claude Code to analyze commits, writ
...
+
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply CHANGELOG.md:45]: Addressed in f867f9433. Relabeled to `[Issue 3104]` while keeping the URL as `/issues/3104` — #3104 is an issue, not a PR. The verbose original entry was already trimmed in a previous push to match the weight of PR #2822's introducing entry; the cross-reference is now restored with the correct label.

File: react_on_rails/spec/react_on_rails/generators/generator_helper_spec.rb [hunk 7]
```diff
@@ -203,29 +203,26 @@ def self.read
...
 
-    it "does not enable rsc-pro mode for standalone --pro" do
-      allow(self).to receive(:options).and_return({ rsc_pro: false, rsc: false, pro: true })
+    it "does not enable Pro or RSC for a plain install" do
+      allow(self).to receive(:options).and_return({ rsc: false, pro: false })
 
-      expect(use_rsc_pro_mode?).to be(false)
-      expect(use_rsc?).to be(false)
-      expect(use_pro?).to be(true)
+      expect(use_rsc?).to be_falsey
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply react_on_rails/spec/react_on_rails/generators/generator_helper_spec.rb]: Skipping. For this specific test, `be_falsey` and `be(false)` are equivalent because `options` is explicitly stubbed with `rsc: false, pro: false` — both matchers pass identically. Leaving as-is.

File: react_on_rails/spec/react_on_rails/generators/generator_helper_spec.rb [hunk 8]
```diff
@@ -203,29 +203,26 @@ def self.read
...
+      expect(use_pro?).to be_falsey
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply react_on_rails/spec/react_on_rails/generators/generator_helper_spec.rb]: Skipping. For this specific test, `be_falsey` and `be(false)` are equivalent because `options` is explicitly stubbed with `rsc: false, pro: false` — both matchers pass identically. Leaving as-is.

**Comment by ihabadham** [comment]
## Address-review summary

Scan scope: full PR history (no prior summary comment).

### Mattered

- **Prerelease note gate too narrow** (claude[bot] ×3, chatgpt-codex-connector[bot]): addressed in f867f9433. Renamed `rsc_prerelease_note` → `prerelease_note` and gated on `prerelease_ror_version?` instead of `use_rsc?` so the note fires whenever the exact pin fires, including for `--pro` on a prerelease gem.
- **Stale method name `rsc_pro_verification_message`** (claude[bot] ×2): addressed in f867f9433. Renamed method + call site to `rsc_verification_message`. Heading text "🔎 RSC Pro Verification" kept — it accurately describes what's being verified.
- **CHANGELOG link label** (coderabbitai[bot] correct, claude[bot] ×1 incorrect): addressed in f867f9433. Labeled as `[Issue 3104]` while keeping the `/issues/3104` URL (3104 is an issue, not a PR). Declined claude[bot]'s suggestion to change the URL to `/pull/3104` since that would be broken.

### Optional

- **Missing `--pro` + prerelease test context** (claude[bot]): declined. Existing `--rsc` + prerelease test covers the `pro_gem_version_requirement` code path; a parallel `--pro` test is defense-in-depth but would largely duplicate coverage.
- **`be_falsey` vs `be(false)` matcher consistency** (claude[bot] ×2): declined. For the test as written (`options` explicitly stubbed with `rsc: false, pro: false`), both matchers behave identically.

### Skipped

- None.

Next default scan starts after this comment. Say `check all reviews` to rescan the full PR.

File: react_on_rails/lib/generators/react_on_rails/pro_setup.rb [hunk 9]
```diff
@@ -60,13 +60,10 @@ def missing_pro_gem?(force: false)
...
 
-        context_line = pro_gem_requirement_context_line
-        prerelease_note = rsc_pro_prerelease_note
-
         GeneratorMessages.add_error(<<~MSG.strip)
           🚫 Failed to auto-install #{PRO_GEM_NAME} gem.
 
-          #{context_line}
+          #{pro_gem_requirement_context_line}
           #{prerelease_note}
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply react_on_rails/lib/generators/react_on_rails/pro_setup.rb]: Addressed in f7c161aff. Introduced `optional_prerelease_line` that is either `""` or a leading newline plus the note, and appended it to the preceding context line so the error message has one blank line regardless of whether the note is present.

File: react_on_rails/spec/react_on_rails/generators/generator_helper_spec.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (ihabadham) [reply react_on_rails/spec/react_on_rails/generators/generator_helper_spec.rb]: Duplicate of a previously-declined style nit from this review cycle. For this specific test, `options` is explicitly stubbed with `rsc: false, pro: false`, so `be_falsey` and `be(false)` behave identically. Leaving as-is.

**Comment by ihabadham** [comment]
Thanks @claude, addressing each point:

**Issue 1 (`--pro` + prerelease test):** declined. The `--rsc` + prerelease test already exercises `pro_gem_version_requirement` and `prerelease_ror_version?` through the same code path; a parallel `--pro` test would largely duplicate that coverage. Unit coverage for `JsDependencyManager` version pinning lives in `js_dependency_manager_spec.rb`.

**Issue 2 (npm version-pin assertions):** addressed in f7c161aff. Strengthened the `--rsc` context's `installs RSC npm dependencies` test from key-presence to exact version-string pins — the same assertions the deleted `--rsc-pro` context carried.

**Issue 3 (friendly migration message for `--rsc-pro`):** skipping. The flag only shipped in 16.6.0 (~2 weeks ago), real-world adoption is negligible, and the generic Thor error is adequate. Deliberate clean break.

**Nits:**
- Double blank line: addressed in f7c161aff (see inline thread).
- `be_falsey` vs `be(false)`: left as-is, see inline thread — matchers are equivalent for this stubbed test.

**Comment by ihabadham** [comment]
## Address-review summary

Scan scope: review activity after 2026-04-21T18:20:41Z (previous summary).

### Mattered

- **Breaking-change UX / migration message for `--rsc-pro`** (claude[bot]): declined. The flag only shipped in 16.6.0 (~2 weeks ago) so adoption is negligible; a clean break with the generic Thor error is adequate.

### Optional

- **Double blank line in `missing_pro_gem?` error message** (claude[bot]): addressed in f7c161aff. Pre-computed `optional_prerelease_line` to avoid the empty-interpolation-on-its-own-line pattern.
- **Port npm version-pin assertions to `--rsc` context** (claude[bot]): addressed in f7c161aff. Strengthened the surviving `--rsc` test from key-presence to exact version-string pins using `ReactOnRails::VersionSyntaxConverter` and `RSC_PACKAGE_VERSION_PIN`.

### Skipped

- **coderabbitai acknowledgment** (3119517748): auto-reply on an already-resolved CHANGELOG thread confirming the fix — no action needed.
- **`be_falsey` vs `be(false)` re-raised** (3119530989): duplicate of a style nit declined earlier this cycle; matchers behave identically for this explicitly-stubbed test. Replied and resolved.
- **Missing `--pro` + prerelease test** (in claude[bot] review 4290857020): duplicate of an optional item declined in the previous cycle. Folded the rationale into the general reply.

Next default scan starts after this comment. Say `check all reviews` to rescan the full PR.

File: react_on_rails/spec/react_on_rails/generators/generator_helper_spec.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (ihabadham) [reply react_on_rails/spec/react_on_rails/generators/generator_helper_spec.rb]: Addressed in ec61e4d57. Checked the broader spec tree first — `be(false)`/`be(true)` is used ~45 times vs `be_falsey` 2 times (both introduced by this PR). `be_falsey` was a clear outlier; aligned with the convention.

**Comment by ihabadham** [comment]
Thanks @claude for the re-review. Addressing each point:

**`be_falsey` inconsistency:** addressed in ec61e4d57. Confirmed via broader pattern scan (`be(false)` used ~45× in spec tree vs `be_falsey` 2× — both were mine from this PR), then aligned to the majority convention.

**`--rsc --pro` combo documentation gap:** skipping. The existing `--rsc` description already says "(includes Pro)", which implicitly tells users `--pro` is redundant with it. Adding an explicit "`--pro` is redundant with `--rsc`" clause is belt-and-suspenders wording for a combination few users will type. The behavior is correct (both flags silently resolve to the same install path); the help text stays concise.

**Comment by ihabadham** [comment]
## Address-review summary

Scan scope: review activity after 2026-04-21T18:45:29Z (previous summary).

### Mattered

- None.

### Optional

- **`be_falsey` → `be(false)` style alignment** (claude[bot]): addressed in ec61e4d57. Verified the broader pattern first — `be(false)`/`be(true)` used ~45× in the spec tree vs `be_falsey` used 2× (both from this PR). Aligned to majority convention; `be(false)` is also stricter (fails on unintended `nil` returns).
- **`--rsc --pro` combo documentation gap** (claude[bot] review 4290996603): declined. Existing `--rsc` description already says "(includes Pro)", implicitly conveying that `--pro` is redundant with it. Adding an explicit "`--pro` is redundant with `--rsc`" clause is belt-and-suspenders for a combination few users will type.

### Skipped

- None.

Next default scan starts after this comment. Say `check all reviews` to rescan the full PR.

File: react_on_rails/lib/generators/react_on_rails/pro_setup.rb [hunk 10]
```diff
@@ -537,13 +531,20 @@ def pro_gem_auto_install_command
       end
 
       def pro_gem_version_requirement
-        # RSC Pro uses exact pinning so the Pro gem version always matches the
-        # paired RSC package version generated in the same run.
-        return ReactOnRails::VERSION if use_rsc_pro_mode?
+        # Prerelease gem versions need an exact pin: Bundler's pessimistic operator
+        # (~>) does not match prerelease versions, so a stable range would fail to
+        # install during prerelease cycles.
+        return ReactOnRails::VERSION if prerelease_ror_version?
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply react_on_rails/lib/generators/react_on_rails/pro_setup.rb:537]: Declining. The behavior is gated on `prerelease_ror_version?`, not on the flag — a `--pro` + prerelease test would exercise the same conditional branch as the existing `--rsc` + prerelease test. The flag is irrelevant to what the gate checks. Accepting the coverage gap for velocity.

**Comment by ihabadham** [comment]
Thanks @claude. Going through each point:

**PR description accuracy** — skipping. The behavior is correct (exact pin fires on prerelease regardless of flag); the description wording is slightly narrower than reality but not misleading.

**Missing `--pro` + prerelease test** — declined (see inline thread). Same code path as the existing `--rsc` prerelease test since the gate is version-based, not flag-based.

**Missing stable-version pinning test** — skipping. Tests the opposite branch of `pro_gem_version_requirement`, which would be genuine coverage. Accepting the gap; a future regression on the `~>` branch would surface quickly since it's the default case for any Pro install on a stable version.

**`recovery_install_command` note** — no action needed, you noted the intent was already covered by `expect(command).not_to include("--pro")`.

**`guavapass.com` exclusion** — already verified earlier in this cycle. Site responds HTTP 200 from non-CI hosts; the TLS handshake failure is specific to GitHub Actions IPs (bot filter). The reference in PROJECTS.md is to a legitimate customer, so the exclusion is the correct fix over removing the link.

**Comment by ihabadham** [comment]
## Address-review summary

Scan scope: review activity after 2026-04-21T18:59:05Z (previous summary).

### Mattered

- None.

### Optional

- **PR description accuracy** (claude[bot]): declined. Behavior is correct (exact pin fires on prerelease regardless of flag); description wording is slightly narrower than reality but not misleading.
- **Missing `--pro` + prerelease test** (claude[bot]): declined. Third raising of this. The gate is `prerelease_ror_version?` (version-based, not flag-based), so a `--pro` + prerelease test exercises the same conditional as the existing `--rsc` + prerelease test.
- **Missing stable-version pinning test** (claude[bot]): declined. Would test the opposite branch of `pro_gem_version_requirement`; real coverage gap but accepted for velocity — the `~>` branch is the default case so a regression would surface quickly in practice.

### Skipped

- **`recovery_install_command` negative assertion** (claude[bot]): non-actionable; reviewer confirmed the intent was already covered.
- **`guavapass.com` exclusion verification** (claude[bot]): already verified earlier — site responds HTTP 200 from non-CI hosts; TLS handshake fails only from GitHub Actions IP range (bot filter). Exclusion is the correct fix for a legitimate PROJECTS.md customer reference.

Next default scan starts after this comment. Say `check all reviews` to rescan the full PR.
